File: backend/app/api/routers/admin/sequences.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_admin, get_db
from app.models.admin_user import AdminUser
from app.repositories import sequences as seq_repo
from app.schemas.sequence import (
    SequenceCreate,
    SequenceOut,
    SequenceStepCreate,
    SequenceStepOut,
    SequenceUpdate,
)

router = APIRouter()


@router.get("", response_model=list[SequenceOut])
async def list_sequences(
    session: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_admin),
) -> list[SequenceOut]:
    items = await seq_repo.list_sequences(session)
    return [SequenceOut.model_validate(s) for s in items]


@router.post("", response_model=SequenceOut, status_code=status.HTTP_201_CREATED)
async def create_sequence(
    body: SequenceCreate,
    session: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_admin),
) -> SequenceOut:
    seq = await seq_repo.create(session, **body.model_dump())
    return SequenceOut.model_validate(seq)


@router.get("/{sequence_id}", response_model=SequenceOut)
async def get_sequence(
    sequence_id: UUID,
    session: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_admin),
) -> SequenceOut:
    seq = await seq_repo.get_by_id(session, sequence_id)
    if not seq:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Sequence not found")
    return SequenceOut.model_validate(seq)


@router.patch("/{sequence_id}", response_model=SequenceOut)
async def update_sequence(
    sequence_id: UUID,
    body: SequenceUpdate,
    session: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_admin),
) -> SequenceOut:
    seq = await seq_repo.get_by_id(session, sequence_id)
    if not seq:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Sequence not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(seq, field, value)
    await session.flush()
    await session.refresh(seq)
    return SequenceOut.model_validate(seq)


@router.get("/{sequence_id}/steps", response_model=list[SequenceStepOut])
async def list_steps(
    sequence_id: UUID,
    session: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_admin),
) -> list[SequenceStepOut]:
    if not await seq_repo.get_by_id(session, sequence_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Sequence not found")
    steps = await seq_repo.list_steps(session, sequence_id)
    return [SequenceStepOut.model_validate(s) for s in steps]


@router.post(
    "/{sequence_id}/steps",
    response_model=SequenceStepOut,
    status_code=status.HTTP_201_CREATED,
)
async def add_step(
    sequence_id: UUID,
    body: SequenceStepCreate,
    session: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_admin),
) -> SequenceStepOut:
    if not await seq_repo.get_by_id(session, sequence_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Sequence not found")
    step = await seq_repo.add_step(
        session,
        sequence_id=sequence_id,
        position=body.position,
        delay_minutes=body.delay_minutes,
        material_id=body.material_id,
    )
    return SequenceStepOut.model_validate(step)


@router.delete("/{sequence_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_sequence(
    sequence_id: UUID,
    session: AsyncSession = Depends(get_db),
    _: AdminUser = Depends(get_current_admin),
) -> None:
    seq = await seq_repo.get_by_id(session, sequence_id)
    if not seq:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Sequence not found")
    await session.delete(seq)

File: frontend/src/api/sequences.ts
import { http } from './client';
import type { Sequence, SequenceStep, TriggerKind } from '../types';

export interface SequencePayload {
  name: string;
  description?: string | null;
  trigger_kind?: TriggerKind;
  is_active?: boolean;
}

export interface StepPayload {
  position: number;
  delay_minutes: number;
  material_id: string;
}

export const sequenceApi = {
  list: () => http.get<Sequence[]>('/api/admin/sequences'),
  get: (id: string) => http.get<Sequence>(`/api/admin/sequences/${id}`),
  create: (data: SequencePayload) =>
    http.post<Sequence>('/api/admin/sequences', data),
  update: (id: string, data: Partial<SequencePayload>) =>
    http.patch<Sequence>(`/api/admin/sequences/${id}`, data),

  remove: (id: string) => http.delete(`/api/admin/sequences/${id}`),

  listSteps: (id: string) =>
    http.get<SequenceStep[]>(`/api/admin/sequences/${id}/steps`),
  addStep: (id: string, data: StepPayload) =>
    http.post<SequenceStep>(`/api/admin/sequences/${id}/steps`, data),
  updateStep: (stepId: string, data: Partial<StepPayload>) =>
    http.patch<SequenceStep>(`/api/admin/sequence-steps/${stepId}`, data),
  deleteStep: (stepId: string) =>
    http.delete(`/api/admin/sequence-steps/${stepId}`),
};
